### PR TITLE
[+] IP Whitelist - Add IP Whitelist feature

### PR DESCRIPTION
**Pull request by ArnaudValensi**
## On express

The IP Whitelist is retrieved:
- On server starts
- On login
- On IP whitelist middleware when:
  - IP Whitelist was not retrieved
  - Or IP not valid in the current whitelist

The IP is checked in the middleware

The errors are:
`httpError(403, 'IP address rejected (' + ip + ')')`
`httpError(403, 'IP whitelist not retrieved')`

## On rails

The retrieving on server start has been removed, since not very useful and difficult to implement.

## To use the IP Whitelist

Use this for the front https://github.com/ForestAdmin/forestadmin/pull/982

## Unit test

`IpWhitelistChecker` is unit tested:
```
$ bundle exec rspec --color --format doc
```

## Pull Request checklist:

- [x] Write an explicit title for the Pull Request
- [x] Write changes made in the CHANGELOG.md
- [x] Create automatic tests
- [x] No automatic tests failures
- [x] Test manually the implemented changes
- [x] Review my own code (indentation, syntax, style, simplicity, readability)
- [x] Wonder if you can improve the existing code

### CODE REVIEW

**Comment by ArnaudValensi** [comment]
[(4) IP Whitelist - Retrieve IPs on start and Add a middleware to check the IP and kick the user (rails)](https://trello.com/c/3WOVHcTh/2410-4-ip-whitelist-retrieve-ips-on-start-and-add-a-middleware-to-check-the-ip-and-kick-the-user-rails)

File: .travis.yml [hunk 1]
```diff
@@ -6,4 +6,4 @@ rvm:
   - 2.2.0
 script:
   - RAILS_ENV=test bundle exec rake --trace db:migrate test
-  - RAILS_ENV=test bundle exec rspec
+  - RAILS_ENV=test bundle exec rspec --color --format doc
```
Comment (ArnaudValensi) [inline .travis.yml]: The details of the tests are not visible without this.

File: app/services/forest_liana/ip_whitelist.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
+module ForestLiana
```
Comment (ArnaudValensi) [inline app/services/forest_liana/ip_whitelist.rb]: Here this is exactly the same logic than on forest-express.

File: app/services/forest_liana/ip_whitelist.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      end
+    end
+
+    def self.is_ip_whitelist_retrieved
+      !@@use_ip_whitelist || !@@ip_whitelist_rules.nil?
+    end
+
+    def self.is_ip_valid(ip)
+      if @@use_ip_whitelist
+        return IpWhitelistChecker.is_ip_matches_any_rule(ip, @@ip_whitelist_rules)
```
Comment (ArnaudValensi) [inline app/services/forest_liana/ip_whitelist.rb]: Except here, some code have been put in IpWhitelistChecker

File: app/controllers/forest_liana/base_controller.rb [hunk 4]
```diff
@@ -2,5 +2,31 @@ module ForestLiana
...
+        if !IpWhitelist.is_ip_whitelist_retrieved || !IpWhitelist.is_ip_valid(ip)
+          unless IpWhitelist.retrieve
+            raise Errors::HTTP403Error.new("IP whitelist not retrieved")
+          end
+
+          unless IpWhitelist.is_ip_valid(ip)
+            raise Errors::HTTP403Error.new("IP address rejected (#{ip})")
+          end
+        end
+      rescue => exception
```
Comment (arnaudbesnier) [inline app/controllers/forest_liana/base_controller.rb]: If you want to rescue `Errors::HTTP403Error` only you can:
`rescue Errors::HTTP403Error => exception`

Other exceptions should be handled differently as `display_error` is not a function implemented for any error object. With another generic `rescue => exception` below the first one.

File: spec/ip_whitelist_checker_spec.rb [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+            'type' => 1,
+            'ip_minimum' => '90.88.0.1',
+            'ip_maximum' => '90.88.255.255',
+          }
+
+          is_matching = IpWhitelistChecker.is_ip_matches_rule('90.88.118.79', rule)
+
+          expect(is_matching).to eq(true)
+        end
+      end
```
Comment (arnaudbesnier) [inline spec/ip_whitelist_checker_spec.rb]: For the 4 tests above a few context is missing, the rule is the same but not the value checked. It should be readable in the tests:

```ruby
describe 'With range rule "90.88.0.0 - 90.88.255.255' do
  rule = {
    'type' => 1,
    'ip_minimum' => '90.88.0.0',
    'ip_maximum' => '90.88.255.255',
  }

  describe 'check X value' do
  end
  
  describe 'check Y value' do
  end

  #... and so on
end
```

File: spec/ip_whitelist_checker_spec.rb [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+          rule = {
+            'type' => 2,
+            'range' => '90.88.0.0/24',
+          }
+
+          is_matching = IpWhitelistChecker.is_ip_matches_rule('90.88.1.1', rule)
+
+          expect(is_matching).to eq(false)
+        end
+      end
```
Comment (arnaudbesnier) [inline spec/ip_whitelist_checker_spec.rb]: Same here for the global structure of the tests.

File: config/initializers/errors.rb [hunk 7]
```diff
@@ -47,8 +53,8 @@ def self.recursively_print(error, margin: '', is_error: false)
           recursively_print(error.cause, margin: "#{margin} ", is_error: is_error)
         else
           is_error ?
-            FOREST_LOGGER.debug(error.backtrace.join("\n")) :
-            FOREST_LOGGER.error(error.backtrace.join("\n"))
+            FOREST_LOGGER.error(error.backtrace.join("\n")) :
+            FOREST_LOGGER.debug(error.backtrace.join("\n"))
```
Comment (arnaudbesnier) [inline config/initializers/errors.rb]: Waiting the rebase to make this disappear?

File: app/services/forest_liana/ip_whitelist.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+      end
+
+      return true
```
Comment (arnaudbesnier) [inline app/services/forest_liana/ip_whitelist.rb]: a simple `true` is enough with Ruby (like you did in the function above)

File: app/controllers/forest_liana/sessions_controller.rb [hunk 9]
```diff
@@ -1,5 +1,7 @@
 module ForestLiana
   class SessionsController < ForestLiana::BaseController
+    skip_before_action :reject_unauthorized_ip
```
Comment (arnaudbesnier) [inline app/controllers/forest_liana/sessions_controller.rb]: Maybe we could remove this skip if its done only for the scenario we discussed.

File: app/controllers/forest_liana/sessions_controller.rb [hunk 10]
```diff
@@ -53,7 +55,7 @@ def process_login(
           raise Errors::HTTP401Error
         end
 
-        # TODO: Add ip whitelist retrieving when it exists.
+        IpWhitelist.retrieve
```
Comment (arnaudbesnier) [inline app/controllers/forest_liana/sessions_controller.rb]: I would add a comment to explain why we do this specific `IpWhitelist.retrieve` here.

**Pull request merged**